### PR TITLE
feat(administration): Helm chart for deploying Vector with an aggregator role

### LINKED ISSUE #3227: Deploy Vector as an aggregator in Kubernetes
[v1](https://github.com/timberio/vector-product/issues/12) of our Kubernetes integration dealt with deploying Vector as a daemon on everyone node to collect app/service logs, but it did not cover deploying Vector as an aggregator (service). We should extend our Helm charts to offer this as an option, as well as updating our documentation.

### PR DESCRIPTION
**Pull request by MOZGIII**
Closes #3227.

This PR implements the design outlined at https://github.com/timberio/vector/issues/3227#issuecomment-701367762. The chart is based on #3840.

To do:
- [x] allow sharing values via `global` - see https://helm.sh/docs/chart_template_guide/subcharts_and_globals/
- [x] E2E tests
- [x] make `Service` `ports` configurable
- [x] make `container` `ports` configurable
- [x] backport the deprecation of the hard-coded config maps from https://github.com/timberio/vector/pull/4938
- [x] backport #5006
- [x] ~ability to disable `vector-agent` or `vector-aggregator` when using `vector` chart~ rejected: people should use individual charts for now if they don't want the full package
- [x] expose more variants of Kuberntes YAML files - `vector-all.yaml`, `vector-agent.yaml` and `vector-aggregator.yaml`, ensure they're correct (#5042 tracks this, so doesn't *have* to happen at this PR - however it makes sense to do the second part at this PR)
- [x] fix clippy at E2E tests
- [ ] docs (covered by #5041 so doesn't need to happen as part of this task)

### CODE REVIEW

File: distribution/helm/vector-agent/values.yaml [hunk 1]
```diff
@@ -108,6 +109,23 @@ kubernetesLogsSource:
   # Raw TOML config to embed at the kubernetes logs source.
   rawConfig: null
 
+# The "built-in" vector sink, to send the logs to the vector aggregator.
+# Disabled by default when using `vector-agent` chart.
+# Enabled by default when usong `vector` chart, and automatically configured.
```
Comment (Hoverbear) [inline distribution/helm/vector-agent/values.yaml:114]:
```suggestion
# Enabled by default when using `vector` chart, and automatically configured.
```

File: lib/k8s-e2e-tests/tests/vector.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      rawConfig: |
+        target = "stdout"
+        encoding = "json"
+"#;
+
+/// This test validates that vector picks up logs with an agent and
+/// delivers them to the aggregator out of the box.
+#[tokio::test]
+async fn logs() -> Result<(), Box<dyn std::error::Error>> {
+    let _guard = lock();
```
Comment (eeyun) [inline lib/k8s-e2e-tests/tests/vector.rs:26]: This is a cool trick. I'm sure we're using it elsewhere in the k8s test code but I just wanted to mention it :)

File: distribution/kubernetes/vector.yaml [hunk 3]
```diff
@@ -189,3 +281,81 @@ spec:
...
+        app.kubernetes.io/instance: vector
+    spec:
+      serviceAccountName: vector-aggregator
+      securityContext:
+        {}
+      containers:
+        - name: vector
+          securityContext:
+            {}
+          image: "timberio/vector:0.0.0-debian"
```
Comment (eeyun) [inline distribution/kubernetes/vector.yaml:319]: Is this intentional?

**Review by eeyun — APPROVED** [review]
Content of this looks good! As an aside - do we have a separate issue to address consumption documentation?

File: distribution/kubernetes/vector.yaml (hunk 3 above)
Comment (MOZGIII) [reply distribution/kubernetes/vector.yaml:319]: It's not, good catch! I've added this to the to-do list in this PR.

**Comment by MOZGIII** [comment]
> do we have a separate issue to address consumption documentation?

I'm not sure yet how do we plan on addressing the Helm chart documentation. We do want to do this though, I've added this to the to-do list.

It might be useful to prepare the documentation systems upfront to allow an easier landing of this after to 0.11 release.

cc @binarylogic

**Comment by spencergilbert** [comment]
@MOZGIII moving over from the other issue... below are my steps and output from using the managed PVC value.

```
☸ dev (vector) in ~
❯ kind create cluster
Creating cluster "kind" ...
 ✓ Ensuring node image (kindest/node:v1.19.1) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-kind"
You can now use your cluster with:

kubectl cluster-info --context kind-kind

Have a question, bug, or feature request? Let us know! https://kind.sigs.k8s.io/#community 🙂

☸ kind-kind in ~/C/v/distribution/helm/vector-aggregator on  helm-aggregator-role is  v0.0.0 via ⎈ v3.4.0
❯ helm template test  . --set storage.mode=managedPersistentVolumeClaim --set storage.managedPersistentVolumeClaim.storageClass=- | kubectl apply -f -
walk.go:74: found symbolic link in path: /home/sgilbert/Code/vector/distribution/helm/vector-aggregator/charts/vector-shared resolves to /home/sgilbert/Code/vector/distribution/helm/vector-shared
serviceaccount/test-vector-aggregator created
configmap/test-vector-aggregator created
service/test-vector-aggregator-headless created
service/test-vector-aggregator created
statefulset.apps/test-vector-aggregator created
```

```
❯ kubectl describe pod test-vector-aggregator-0
Name:           test-vector-aggregator-0
Namespace:      default
Priority:       0
Node:           <none>
Labels:         app.kubernetes.io/instance=test
                app.kubernetes.io/name=vector-aggregator
                controller-revision-hash=test-vector-aggregator-755b7b966
                statefulset.kubernetes.io/pod-name=test-vector-aggregator-0
Annotations:    checksum/config: 69d2bc6fd53d7c164cb3377fe0495bebbec5fb5b0a843b489572d5faf2a3749a
Status:         Pending
IP:             
IPs:            <none>
Controlled By:  StatefulSet/test-vector-aggregator
Containers:
  vector:
    Image:      timberio/vector:0.0.0-debian
    Port:       <none>
    Host Port:  <none>
    Args:
      --config
      /etc/vector/*.toml
    Environment:  <none>
    Mounts:
      /etc/vector from config-dir (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from test-vector-aggregator-token-qcgdd (ro)
      /vector-data-dir from data-dir (rw)
Conditions:
  Type           Status
  PodScheduled   False 
Volumes:
  test-vector-aggregator-data-dir:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  test-vector-aggregator-data-dir-test-vector-aggregator-0
    ReadOnly:   false
  data-dir:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  test-vector-aggregator-data-dir
    ReadOnly:   false
  config-dir:
    Type:                Projected (a volume that contains injected data from multiple sources)
    ConfigMapName:       test-vector-aggregator
    ConfigMapOptional:   0xc000867989
    ConfigMapName:       vector-aggregator-config
    ConfigMapOptional:   0xc00086798a
    SecretName:          vector-aggregator-config
    SecretOptionalName:  0xc00086798b
  test-vector-aggregator-token-qcgdd:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  test-vector-aggregator-token-qcgdd
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node-role.kubernetes.io/master:NoSchedule
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason            Age                From               Message
  ----     ------            ----               ----               -------
  Warning  FailedScheduling  27s (x3 over 90s)  default-scheduler  persistentvolumeclaim "test-vector-aggregator-data-dir" not found
```

**Comment by MOZGIII** [comment]
@spencergilbert I found and fixed the issue, great find! I wonder how I missed this...

**Comment by eeyun** [comment]
Updates since previous review look good @MOZGIII. Let me know when you're ready for a new review :+1:

**Pull request merged**